### PR TITLE
[tbl111] Use pipenv for deterministic builds

### PR DESCRIPTION
**Pull request by benjefferies**
Pipenv enables us to lock down and guarantee what dependencies we get
when we install dependencies. This works by using the pip hashes when
running `pipenv install --dev`. The hashes are stored in a generated
file call Pipfile.lock.

We no longer need requirement files instead pipenv uses a file called
Pipfile. The documentation can be found
https://github.com/kennethreitz/pipenv.

* Bump cryptography library

* Use latest dev dependencies

**What is the context of this PR?**

Link to Trello card

https://trello.com/c/TNrTo9L1/733-tbl111-create-reproducible-deployments

**How to review**

1. Check out locally and follow README.md to set up with pipenv
1. `cf push` to check deployment

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,18 +1,16 @@
 language: python
 sudo: required
 dist: trusty
-python: '3.4'
+python: '3.5'
```
Comment (jwgwalton) [inline .travis.yml]: Why not 3.6?

File: README.md [hunk 2]
```diff
@@ -5,27 +5,21 @@
...
+Install pipenv
+```bash
+pip install pipenv
 ```
 
-Install dependencies for the application using pip
-
+Use pipenv to create a virtualenv and install dependencies
+```bash
+pipenv --python 3.5 install --dev
```
Comment (jwgwalton) [inline README.md]: As above, why not 3.6?

File: Pipfile [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+cryptography = "==2.1.2"
+notifications-python-client = "==4.2.0"
+cfenv = "==0.5.3"
+gunicorn = "==19.7.1"
+
+
+[dev-packages]
+
+pylint = "*"
+"flake8" = "*"
```
Comment (jwgwalton) [inline Pipfile]: Playing devils advocate, if we are not enforcing flake8 in this repo do we need the import?

File: .travis.yml (hunk 1 above)
Comment (benjefferies) [reply .travis.yml]: Probably should be going forward but kept it inline with the runtime.txt for development

File: Pipfile (hunk 3 above)
Comment (benjefferies) [reply Pipfile]: Good point. We probably should be. I'll run it and see if it gives me any gyp

Comment (benjefferies) [reply Pipfile]: We've got some error so going to leave it out for now. Will raise a tech debt ticket

Comment (jwgwalton) [reply Pipfile]: Yeh, I ran a flake8 check in tox and it wasn't pretty. Do we have anyway to lint the code?

Comment (benjefferies) [reply Pipfile]: I guess add it in and fix the linting issues. I'll pick up after the pipenv stuff if I can

File: README.md [hunk 4]
```diff
@@ -5,49 +5,39 @@
...
+Install pipenv
+```bash
+pip install pipenv
 ```
 
-Install dependencies for the application using pip
-
+Use pipenv to create a virtualenv and install dependencies
+```bash
+pipenv run install --dev
```
Comment (jcox-dev) [inline README.md]: pipenv run install -> pipenv install

File: README.md [hunk 5]
```diff
@@ -5,49 +5,39 @@
...
-
-```
-pip install -r requirements-test.txt
-```
-Once these have been installed the tests can be run from the ras-secure-message directory using the following
+[Install Docker](https://docs.docker.com/engine/installation/)
 ```
-python run_tests.py
+docker run -d -p 5432:5432 --name postgres -e POSTGRES_PASSWORD=postgres -e POSTGRES_DB=postgres -d postgres
+pipenv runcoverage run run_tests.py
```
Comment (jcox-dev) [inline README.md]: pipenv runcoverage run run_tests.py -> pipenv run run_tests.py

File: README.md [hunk 6]
```diff
@@ -5,56 +5,58 @@
...
 
 ## Setup
-Based on python 3.4
 
-Create a virtual env for python3 and make sure that it's activated
+[Install Docker](https://docs.docker.com/engine/installation/)
 
+Install pipenv
+```bash
+pip install pipenv
```
Comment (jwgwalton) [inline README.md]: I get the following error? 

Installing dependencies from Pipfile.lock (0ce6d1)…
An error occurred while installing psycopg2==2.6.2! Will try again.
  🐍   ▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉ 38/38 — 00:00:03
Installing initially–failed dependencies…
Collecting psycopg2==2.6.2 ▉▉▉▉▉▉▉▉▉▉ 0/1 — 00:00:00
  Using cached psycopg2-2.6.2.tar.gz
    Complete output from command python setup.py egg_info:
    running egg_info
    creating pip-egg-info/psycopg2.egg-info
    writing pip-egg-info/psycopg2.egg-info/PKG-INFO
    writing dependency_links to pip-egg-info/psycopg2.egg-info/dependency_links.txt
    writing top-level names to pip-egg-info/psycopg2.egg-info/top_level.txt
    writing manifest file 'pip-egg-info/psycopg2.egg-info/SOURCES.txt'
    Error: pg_config executable not found.
    
    Please add the directory containing pg_config to the PATH
    or specify the full executable path with the option:
    
        python setup.py build_ext --pg-config /path/to/pg_config build ...
    
    or with the pg_config option in 'setup.cfg'.
    
    ----------------------------------------

Command "python setup.py egg_info" failed with error code 1 in /private/var/folders/l8/ppxlhm3s58356981xh6xlk880000gp/T/pip-build-ssz2oc7h/psycopg2/

Comment (jwgwalton) [reply README.md]: This comment should be attached to
 ```bash
pipenv install
```

Comment (jcox-dev) [reply README.md]: @jwgwalton `brew install postgresql` / `apt-get install libpq-dev`

Comment (jcox-dev) [reply README.md]: Probably worth updating README with the system dependency.

Comment (jwgwalton) [reply README.md]: It needs an external dependency, I thought it was all self-contained?

Comment (benjefferies) [reply README.md]: http://initd.org/psycopg/docs/install.html#build-prerequisites

File: Pipfile (hunk 3 above)
Comment (jcox-dev) [reply Pipfile]: NB: It would be useful to fail travis builds on linting issues as well.

**Comment by benjefferies** [comment]
Tested in Int

**Pull request merged**